contracts/TokenProxy.sol

```diff
@@ -0,0 +1,159 @@
+pragma solidity ^0.4.19;
+
+contract TransferableToken {
+    function balanceOf(address _addr) public constant returns (uint256);
+    function transfer(address _to, uint256 _value) public returns (bool success);
+}
+
+
+/// The Gate is a temporary contract with unique address to allow a token holder
+/// (called "User") to transfer token from original Token to the Proxy.
+///
+/// TODO: Rename to MigrationGate?
+contract Gate {
+    TransferableToken private TOKEN;
+    TokenProxy private PROXY;
+    address private USER;
+
+    /// Gates are to be created by the TokenProxy.
+    function Gate(TransferableToken _token, TokenProxy _proxy, address _user) public {
+        TOKEN = _token;
+        PROXY = _proxy;
+        USER = _user;
```

Comment:
We could avoid USER field if we modify close() and onGateClosed() functions: remove require from close(), pass msg.sender to onGateClosed() as a USER parameter, reverse gates mapping as user->gate, and add require(msg.sender == gate) in the onGateClosed() function.

Reply:
Yes, will be done.